crates/file/src/blob.rs

```diff
@@ -0,0 +1,338 @@
+use std::{ops::Deref, time::Duration};
+
+use wasm_bindgen::{prelude::*, JsCast};
+
+mod sealed {
+    pub trait Sealed {}
+}
+use sealed::Sealed;
+
+/// This trait is used to overload the `Blob::new_with_options` function, allowing a variety of
+/// types to be used to create a `Blob`. Ignore this, and use &[u8], &str, etc to create a `Blob`.
+///
+/// The trait is sealed: it must only be implemented by types in this
+/// crate, as this crate relies on invariants regarding the `JsValue` returned from `into_jsvalue`.
+pub trait BlobContents: Sealed {
+    fn into_jsvalue(self) -> JsValue;
+}
+
+impl<'a> Sealed for &'a str {}
+impl<'a> BlobContents for &'a str {
+    fn into_jsvalue(self) -> JsValue {
+        JsValue::from_str(self)
+    }
+}
+
+impl<'a> Sealed for &'a [u8] {}
+impl<'a> BlobContents for &'a [u8] {
+    fn into_jsvalue(self) -> JsValue {
+        let array = unsafe { js_sys::Uint8Array::view(self) };
+        js_sys::Uint8Array::new(&array).into()
+    }
+}
+
+impl Sealed for js_sys::ArrayBuffer {}
+impl BlobContents for js_sys::ArrayBuffer {
+    fn into_jsvalue(self) -> JsValue {
+        self.into()
+    }
+}
+
+/// A [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob).
+///
+/// `Blob`s can be created directly from `&str`, `&[u8]`, and `js_sys::ArrayBuffer`s using the
+/// `From::from` implementation (e.g. `let b: Blob = "test".into()`), or using the
+/// `new_with_options` constructor.
+#[derive(Debug, Clone)]
+pub struct Blob {
+    inner: web_sys::Blob,
+}
+
+impl Blob {
+    fn new(val: JsValue, mime_type: Option<&str>) -> Blob {
+        let mut properties = web_sys::BlobPropertyBag::new();
+        if let Some(mime_type) = mime_type {
+            properties.type_(mime_type);
+        }
+
+        let parts = js_sys::Array::of1(&val);
+        let inner = web_sys::Blob::new_with_u8_array_sequence_and_options(&parts, &properties);
+
+        Blob::from(inner.unwrap_throw())
+    }
+
+    /// Like `new`, but allows specifying the media type (also known as *content type* or *mime
```

Comment:
```suggestion
    /// Like `new`, but allows specifying the MIME type (also known as *content type*) of the `Blob`.
```

Reply:
I spent a bit of time trying to come up with the correct name for mime/media/content type. I think I went for media based on [its wikipedia page](https://en.wikipedia.org/wiki/Media_type) saying that it's now called media. However MIME might be better if it's more recognisable. Which would you prefer?

Reply:
MIME type is used in the W3C spec and MDN, I think it's *far* more familiar and commonly used than media type.

And since the property is called `mime`, I think it should put MIME first, but I don't have a particularly strong opinion on it.

Reply:
What you suggest sounds good :).